### PR TITLE
Use normalized semantic version in package path/name

### PR DESCRIPTION
**Pull request by troydai**
Addressing #2192 

Still running DNX testing at this point.

### CODE REVIEW

File: src/Microsoft.Framework.Runtime/NuGet/SemanticVersion.cs [hunk 1]
```diff
@@ -79,6 +79,21 @@ public string SpecialVersion
 
         public string OriginalString { get; }
 
+        public string GetNormalizedVersionString()
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/NuGet/SemanticVersion.cs]: This was copied from NuGet right?

File: src/Microsoft.Framework.Runtime/NuGet/Packages/DefaultPackagePathResolver.cs [hunk 2]
```diff
@@ -64,7 +64,7 @@ public virtual string GetPackageDirectory(string packageId, SemanticVersion vers
             string directory = packageId;
             if (_useSideBySidePaths)
             {
-                directory = Path.Combine(directory, version.ToString());
```
Comment (troydai) [inline src/Microsoft.Framework.Runtime/NuGet/Packages/DefaultPackagePathResolver.cs]: I'll try to eliminate the usage of `ToString` in scenarios other than debugging.

File: src/Microsoft.Framework.Runtime/NuGet/SemanticVersion.cs [hunk 3]
```diff
@@ -79,6 +79,21 @@ public string SpecialVersion
...
+        {
+            var revision = Version.Revision > 0 ? ("." + Version.Revision.ToString(CultureInfo.InvariantCulture)) : string.Empty;
+            var specialVer = !string.IsNullOrEmpty(SpecialVersion) ? ("-" + SpecialVersion) : string.Empty;
+
+            // SemanticVersion normalizes the missing components to 0.
+            return string.Format(CultureInfo.InvariantCulture,
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/NuGet/SemanticVersion.cs]: Any reason you don't use string interpolation here?

File: src/Microsoft.Framework.Runtime/NuGet/SemanticVersion.cs (hunk 1 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime/NuGet/SemanticVersion.cs]: Yes, it is.

File: src/Microsoft.Framework.Runtime/NuGet/SemanticVersion.cs (hunk 3 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime/NuGet/SemanticVersion.cs]: I believe we can. No particular reason not to.

**Comment by davidfowl** [comment]
Where is the change that always writes the normalized version folder? Does this code affect both reads and writes? Does it work with v2 and v3 feed formats? Local and remote?

**Comment by troydai** [comment]
_Where is the change that always writes the normalized version folder?_
-It is done implicitly. The `NuGetPackageUtils.InstallFromStream` resolve the target path using updatede code. https://github.com/aspnet/dnx/blob/dev/src/Microsoft.Framework.PackageManager/Utils/NuGetPackageUtils.cs#L27. 

_Does this code affect both reads and writes?_
- Base on the code I read so far, yes. Places where NuGet package paths are required, `IPackagePathResolver` is used to resolve the paths. I'll consult @ChengTian and @anurse tomorrow.

_Does it work with v2 and v3 feed formats?_
- Yes, in terms of the scenarios of #2192 . There may be broader scenarios be impacted, I'll seek advices from @ChengTian and @anurse and test tomorrow.

_ Local and remote?_
- As far as I can see, `DefaultPackagePathResolver` is the where the paths are resolved. I'll take a deeper look tomorrow to see if there is other functions resolving the path. If so these functions should be replaced by `IPackagePathResolver`

**Comment by davidfowl** [comment]
Sounds great! I think this fix will work well then. Please add a basic test

**Comment by anurse** [comment]
Looks good to me, agreed a test would be good.

**Comment by ChengTian** [comment]
Yes,all package install paths should be resolved with `DefaultPackagePathResolver`. Looks good to me.

**Comment by troydai** [comment]
Adding a few tests for `SemanticVersion`

**Comment by davidfowl** [comment]
Add a restore test.

**Comment by troydai** [comment]
We don't have test infrastructure for and actual dun restore test yet. Size of the change seems not to fit in this PR. I file this issue to track the work: https://github.com/aspnet/dnx/issues/2231.

**Comment by troydai** [comment]
Ping :watch:

**Comment by davidfowl** [comment]
:shipit:

**Pull request merged**